### PR TITLE
Added [SecurityHeaders] badge

### PR DESCRIPTION
**Pull request by srtab**
Pull request with #3929 contribution.

### CODE REVIEW

**Comment by PyvesB** [comment]
Thanks for your contribution @srtab!

I've just edited the title and kicked the CI so that the new service tests get run as well.

File: services/security-headers/security-headers.service.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
+'use strict'
+
+const { BaseService } = require('..')
+
+const documentation = `
+<p>
+  The <a href="https://securityheaders.com/">Security Headers</a>
+  provide a easy mechanism to analyze HTTP response headers and give information on how to deploy missing headers.
```
Comment (PyvesB) [inline services/security-headers/security-headers.service.js]: Typo: _an_ easy.

File: services/security-headers/security-headers.service.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+          grade: 'A+',
+        }),
+        documentation,
+      },
+    ]
+  }
+
+  static get defaultBadgeData() {
+    return {
+      label: 'securityheaders',
```
Comment (PyvesB) [inline services/security-headers/security-headers.service.js]: I'm wondering whether the badge label should be in two words as well.

File: services/security-headers/security-headers.service.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+module.exports = class SecurityHeaders extends BaseService {
+  static get category() {
+    return 'monitoring'
+  }
+
+  static get route() {
+    return {
+      base: 'security-headers',
+      pattern: ':protocol(https|http)/:host',
```
Comment (PyvesB) [inline services/security-headers/security-headers.service.js]: I would suggest using a query parameter instead, like I did for the [Bugzilla service](https://github.com/badges/shields/blob/master/services/bugzilla/bugzilla.service.js). It has become our preferred approach for new badges. 😉

File: services/security-headers/security-headers.service.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      color: colorMap[grade],
+    }
+  }
+
+  async handle({ protocol, host }) {
+    let grade
+
+    try {
+      const { res } = await this._request({
+        url: `https://securityheaders.com/?q=${protocol}%3A%2F%2F${host}&hide=on&followRedirects=on`,
```
Comment (PyvesB) [inline services/security-headers/security-headers.service.js]: You can use `qs` options like what is done in the [Wercker service](https://github.com/badges/shields/blob/b3be4d94d5ef570b8daccfd088c343a958988843/services/wercker/wercker.service.js#L87). Like that, no issues with URL encoding!

File: services/security-headers/security-headers.service.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        options: {
+          method: 'HEAD',
+        },
+      })
+
+      grade = res.headers['x-grade']
+    } catch (e) {
```
Comment (PyvesB) [inline services/security-headers/security-headers.service.js]: No need to handle errors explicitly here: they would be handled by the base classes in a consistent way!

**Comment by srtab** [comment]
Thanks for the review @PyvesB, i will make the changes as soon as possible.

File: services/security-headers/security-headers.service.js (hunk 3 above)
Comment (srtab) [reply services/security-headers/security-headers.service.js]: Can i define an empty pattern?

File: services/security-headers/security-headers.service.js (hunk 2 above)
Comment (srtab) [reply services/security-headers/security-headers.service.js]: Switched to "security headers".

File: services/security-headers/security-headers.service.js (hunk 4 above)
Comment (srtab) [reply services/security-headers/security-headers.service.js]: Changed to use `qs` options :)

File: services/security-headers/security-headers.service.js (hunk 5 above)
Comment (srtab) [reply services/security-headers/security-headers.service.js]: Removed error handling.

File: services/security-headers/security-headers.service.js (hunk 3 above)
Comment (PyvesB) [reply services/security-headers/security-headers.service.js]: Sorry, probably wasn't very clear. In the Bugzilla service, `baseUrl` is not only used for tests: it's used in production to select the Bugzilla host you're interested in, for instance: https://img.shields.io/bugzilla/545425?baseUrl=https%3A%2F%2Fbugs.eclipse.org%2Fbugs

So what I had in mind was something along the lines of: https://img.shields.io/security-headers?domain=https%3A%2F%2Fshields.io

@paulmelnikow do you have a strong opinion on this one?

File: services/security-headers/security-headers.tester.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+t.create('grade of http://shields.io')
+  .get('/http/shields.io.json')
+  .expectBadge({ label: 'security headers', message: 'F', color: 'red' })
+
+t.create('grade of https://shields.io')
+  .get('/https/shields.io.json')
+  .expectBadge({ label: 'security headers', message: 'F', color: 'red' })
+
+t.create('grade when no network')
```
Comment (PyvesB) [inline services/security-headers/security-headers.tester.js]: This test can now go away as it's behaviour that's already tested within `BaseService`.

File: services/security-headers/security-headers.service.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        method: 'HEAD',
+        qs: {
+          q: `${protocol}://${host}`,
+          hide: 'on',
+          followRedirects: 'on',
+        },
+      },
+    })
+
+    const grade = res.headers['x-grade']
```
Comment (PyvesB) [inline services/security-headers/security-headers.service.js]: Is it worth checking whether grade is defined? Something along the lines of:
````
if (!grade) {
  throw new NotFound({ prettyMessage: 'not available' }
}

````

File: services/security-headers/security-headers.service.js (hunk 3 above)
Comment (calebcartwright) [reply services/security-headers/security-headers.service.js]: IMO we should use a queryParam here. We decided to leverage queryParams vs. route params to take in a user-provided url input to simplify the patterns, and we should be consistent here. 

There's probably still a couple of existing services that haven't been updated yet, but for a net new service like this one we should follow the queryParam approach. 

See https://github.com/badges/shields/pull/3633#discussion_r299231749 for more info.

Comment (srtab) [reply services/security-headers/security-headers.service.js]: @PyvesB Is it possible to create a service without `pattern`? I tried with empty `pattern` on route, and `namedParams: {}` on examples but i obtain this error: "Error: Example for SecurityHeaders at index 0 does not declare a pattern". Can you help me with this?

File: services/security-headers/security-headers.tester.js (hunk 6 above)
Comment (srtab) [reply services/security-headers/security-headers.tester.js]: Sure, completely agree!

File: services/security-headers/security-headers.service.js (hunk 3 above)
Comment (calebcartwright) [reply services/security-headers/security-headers.service.js]: That's a good question. I've added some inline suggestions you can make that should get it running again

File: services/security-headers/security-headers.service.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
+'use strict'
+
+const { BaseService } = require('..')
```
Comment (calebcartwright) [inline services/security-headers/security-headers.service.js]:
```suggestion
const Joi = require('@hapi/joi')
const { optionalUrl } = require('../validators')
const { BaseService } = require('..')

const queryParamSchema = Joi.object({
  baseUrl: optionalUrl.required(),
}).required()
```

File: services/security-headers/security-headers.service.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+`
+
+module.exports = class SecurityHeaders extends BaseService {
+  static get category() {
+    return 'monitoring'
+  }
+
+  static get route() {
+    return {
+      base: 'security-headers',
```
Comment (calebcartwright) [inline services/security-headers/security-headers.service.js]:
```suggestion
      base: '',
```

File: services/security-headers/security-headers.service.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      pattern: ':protocol(https|http)/:host',
```
Comment (calebcartwright) [inline services/security-headers/security-headers.service.js]:
```suggestion
      pattern: 'security-headers',
      queryParamSchema,
```

File: services/security-headers/security-headers.service.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    }
+  }
+
+  static get examples() {
+    return [
+      {
+        title: 'Security Headers',
+        namedParams: {
```
Comment (calebcartwright) [inline services/security-headers/security-headers.service.js]:
```suggestion
```

File: services/security-headers/security-headers.service.js [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+          protocol: 'https',
```
Comment (calebcartwright) [inline services/security-headers/security-headers.service.js]:
```suggestion
```

File: services/security-headers/security-headers.service.js [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+          host: 'shields.io',
```
Comment (calebcartwright) [inline services/security-headers/security-headers.service.js]:
```suggestion
```

File: services/security-headers/security-headers.service.js [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+        },
```
Comment (calebcartwright) [inline services/security-headers/security-headers.service.js]:
```suggestion
        namedParams: {},
        queryParams: { baseUrl: 'https://shields.io' },
```

File: services/security-headers/security-headers.service.js (hunk 11 above)
Comment (calebcartwright) [reply services/security-headers/security-headers.service.js]: AFAIK GitHub does not support suggested changes on multi-line blocks, so all of these are basically to convert what you currently have for `namedParams` to `namedParams: {},`

File: services/security-headers/security-headers.service.js (hunk 3 above)
Comment (srtab) [reply services/security-headers/security-headers.service.js]: @calebcartwright Thanks for the suggestions, it works just fine! I already made the suggested changes.

File: services/security-headers/security-headers.tester.js (hunk 6 above)
Comment (srtab) [reply services/security-headers/security-headers.tester.js]: Done, i removed the test.

File: services/security-headers/security-headers.service.js (hunk 7 above)
Comment (srtab) [reply services/security-headers/security-headers.service.js]: I agree. I added the suggested code.

**Comment by PyvesB** [comment]
@srtab do you have a Twitter handle I can give credit to in a Tweet about this new badge?

**Comment by srtab** [comment]
> @srtab do you have a Twitter handle I can give credit to in a Tweet about this new badge?

Sorry for late response. I do have an account, but i haven't used it. But thanks anyway.

**Comment by PyvesB** [comment]
Tweet posted: https://twitter.com/Shields_io/status/1174037594851688448

**Pull request merged**